src/pyoak/match/xpath.py

```diff
@@ -192,4 +223,33 @@ def __init__(self, xpath: str) -> None:
             raise ASTXpathDefinitionError("Incorrect xpath definition") from e
 
     def match(self, node: ASTNode) -> bool:
-        return _match_node_xpath(node, self._elemetns)
+        """Match the `node` to the xpath."""
+        return _match_node_xpath(node, self._elements_reversed)
+
+    def findall(self, root: ASTNode) -> Generator[ASTNode, None, None]:
+        """Find all nodes in the `root` that match the xpath.
+
+        Adapted from antlr4-python3-runtime/src/Python3/antlr4/xpath/Xpath.py
+        Copyright (c) 2012-2017 The ANTLR Project. All rights reserved.
+        Use of this file is governed by the BSD 3-clause license that
+        can be found in the LICENSE.txt file in the project root.
+
+        """
+        work: list[ASTNode] = [_DUMMY_XPATH_ROOT(root, origin=NO_ORIGIN)]
+
+        for el in self._elements:
+            new_work: list[ASTNode] = []
+
+            for node in work:
+                if el.anywhere:
+                    new_work.extend(
```

Comment:
afaik (based on ANTLRs implementation comments that this function is borrowing from) use of a "set" (ordered dict in v2) was important. The same nodes can pop-up again and again due do dfs call, so using a regular list will degrade performance exponentially

Reply:
fixed